### PR TITLE
Fix trim empty string

### LINKED ISSUE #67792: MySQL Compatibility: TRIM does not support empty string as trim_character
`TRIM(LEADING '' FROM 'foo')` results in an error.

When using the [SPLIT function](https://help.tableau.com/current/pro/desktop/en-us/functions_functions_string.htm#split), Tableau Cloud generates a query like the following:
```sql
SELECT TRIM(TRIM(LEADING '_' FROM TRIM(LEADING SUBSTRING_INDEX(`TEST_TABLE`.`TEST_COLUMN`, '_', (1 - 1)) FROM SUBSTRING_INDEX(`TEST_TABLE`.`TEST_COLUMN`, '_', 1)))) AS `TEST_COLUMN - 分割済み 1`
FROM `TEST_TABLE`
GROUP BY 1
ORDER BY `TEST_COLUMN - 分割済み 1` ASC;
```
However, this results in the following error:
```
Code: 36. DB::Exception: The pattern argument is not a valid re2 pattern: missing ]: []+: while executing 'FUNCTION replaceRegexpOne(substringIndex(__table1.TEST_COLUMN, '_'_String, 1_UInt8) :: 0, concat('^['_String, regexpQuoteMeta(substringIndex(__table1.TEST_COLUMN, '_'_String, _CAST(0_Int16, 'Int16'_String))), ']+'_String) :: 2, ''_String : 6) -> replaceRegexpOne(substringIndex(__table1.TEST_COLUMN, '_'_String, 1_UInt8), concat('^['_String, regexpQuoteMeta(substringIndex(__table1.TEST_COLUMN, '_'_String, _CAST(0_Int16, 'Int16'_String))), ']+'_String), ''_String) String : 5'. (BAD_ARGUMENTS)
```
**Expected behavior**
Like MySQL, `TRIM(LEADING '' FROM 'foo')` should not result in an error.

**Error message and/or stacktrace**
Stack Trace:
```
0. DB::Exception::Exception(DB::Exception::MessageMasked&&, int, bool) @ 0x000000000d7a923b
1. DB::Exception::Exception(PreformattedMessage&&, int) @ 0x0000000007b64c0c
2. DB::Exception::Exception<String const&>(int, FormatStringHelperImpl<std::type_identity<String const&>::type>, String const&) @ 0x0000000007b771ab
3. DB::FunctionStringReplace<DB::ReplaceRegexpImpl<DB::(anonymous namespace)::NameReplaceRegexpOne, (DB::ReplaceRegexpTraits::Replace)0>, DB::(anonymous namespace)::NameReplaceRegexpOne>::executeImpl(std::vector<DB::ColumnWithTypeAndName, std::allocator<DB::ColumnWithTypeAndName>> const&, std::shared_ptr<DB::IDataType const> const&, unsigned long) const @ 0x000000000c8ef858
4. DB::FunctionToExecutableFunctionAdaptor::executeImpl(std::vector<DB::ColumnWithTypeAndName, std::allocator<DB::ColumnWithTypeAndName>> const&, std::shared_ptr<DB::IDataType const> const&, unsigned long) const @ 0x0000000007b7c73a
5. DB::IExecutableFunction::executeWithoutLowCardinalityColumns(std::vector<DB::ColumnWithTypeAndName, std::allocator<DB::ColumnWithTypeAndName>> const&, std::shared_ptr<DB::IDataType const> const&, unsigned long, bool) const @ 0x00000000087d90e7
6. DB::IExecutableFunction::executeWithoutSparseColumns(std::vector<DB::ColumnWithTypeAndName, std::allocator<DB::ColumnWithTypeAndName>> const&, std::shared_ptr<DB::IDataType const> const&, unsigned long, bool) const @ 0x00000000087d99fe
7. DB::IExecutableFunction::execute(std::vector<DB::ColumnWithTypeAndName, std::allocator<DB::ColumnWithTypeAndName>> const&, std::shared_ptr<DB::IDataType const> const&, unsigned long, bool) const @ 0x00000000087dad7d
8. DB::ExpressionActions::execute(DB::Block&, unsigned long&, bool, bool) const @ 0x00000000111fe0d9
9. DB::ExpressionTransform::transform(DB::Chunk&) @ 0x00000000130794f6
10. DB::ISimpleTransform::transform(DB::Chunk&, DB::Chunk&) @ 0x000000000dab4573
11. DB::ISimpleTransform::work() @ 0x0000000012e107b2
12. DB::ExecutionThreadContext::executeTask() @ 0x0000000012e2bea7
13. DB::PipelineExecutor::executeStepImpl(unsigned long, std::atomic<bool>*) @ 0x0000000012e20610
14. void std::__function::__policy_invoker<void ()>::__call_impl<std::__function::__default_alloc_func<DB::PipelineExecutor::spawnThreads()::$_0, void ()>>(std::__function::__policy_storage const*) @ 0x0000000012e21cce
15. ThreadPoolImpl<ThreadFromGlobalPoolImpl<false, true>>::worker(std::__list_iterator<ThreadFromGlobalPoolImpl<false, true>, void*>) @ 0x000000000d8670a9
16. void std::__function::__policy_invoker<void ()>::__call_impl<std::__function::__default_alloc_func<ThreadFromGlobalPoolImpl<false, true>::ThreadFromGlobalPoolImpl<void ThreadPoolImpl<ThreadFromGlobalPoolImpl<false, true>>::scheduleImpl<void>(std::function<void ()>, Priority, std::optional<unsigned long>, bool)::'lambda0'()>(void&&)::'lambda'(), void ()>>(std::__function::__policy_storage const*) @ 0x000000000d86b1f1
17. void* std::__thread_proxy[abi:v15007]<std::tuple<std::unique_ptr<std::__thread_struct, std::default_delete<std::__thread_struct>>, void ThreadPoolImpl<std::thread>::scheduleImpl<void>(std::function<void ()>, Priority, std::optional<unsigned long>, bool)::'lambda0'()>>(void*) @ 0x000000000d869f69
18. ? @ 0x000076cb917d0609
19. ? @ 0x000076cb916f5353
```

**Additional context**
* version: ClickHouse 24.8.1.1347
* hash: 1e194596106c727ab3731cc74eab6759969f275d

### PR DESCRIPTION
**Pull request by petern48**
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Allow `TRIM` -ing `LEADING` or `TRAILING` empty string as a no-op. Closes #67792.

### CODE REVIEW

File: src/Parsers/ExpressionListParsers.cpp [hunk 1]
```diff
@@ -1704,14 +1704,16 @@ class TrimLayer : public Layer
                             pattern_list_args->children = {
                                 std::make_shared<ASTLiteral>("^["),
                                 to_remove,
-                                std::make_shared<ASTLiteral>("]+")
+                                /// |[] for to avoid exception in '' case
+                                std::make_shared<ASTLiteral>("]+|[]")
```
Comment (alexey-milovidov) [inline src/Parsers/ExpressionListParsers.cpp:1708]: The empty character set looks wrong.

File: tests/queries/0_stateless/03224_trim_empty_string.sql [hunk 2]
```diff
@@ -0,0 +1,3 @@
+SELECT trim(LEADING '' FROM 'foo');
```
Comment (alexey-milovidov) [inline tests/queries/0_stateless/03224_trim_empty_string.sql:1]: This wasn't a bug, but merely an implementation choice.

File: src/Parsers/ExpressionListParsers.cpp [hunk 3]
```diff
@@ -1704,14 +1704,16 @@ class TrimLayer : public Layer
                             pattern_list_args->children = {
                                 std::make_shared<ASTLiteral>("^["),
                                 to_remove,
-                                std::make_shared<ASTLiteral>("]+")
+                                /// |[] for to avoid exception in '' case
```
Comment (alexey-milovidov) [inline src/Parsers/ExpressionListParsers.cpp:1707]: Implementing it in the parser and with regular expressions is a strange choice.

File: tests/queries/0_stateless/03224_trim_empty_string.sql (hunk 2 above)
Comment (petern48) [reply tests/queries/0_stateless/03224_trim_empty_string.sql:1]: I see. So we should keep the current behavior and let it raise an exception? In that case, I guess we should close this PR and the original issue.

**Comment by alexey-milovidov** [comment]
It makes sense: MySQL, Postgres, and MS SQL support this case.

**Comment by petern48** [comment]
I was trying to keep the change self-contained, but I agree the regex method was weird. I removed that commit and made it so the code skips the replaceRegexp function entirely for this case. Implementing in the parser still felt the natural to avoid passing an invalid regexp down the line, but let me know if there's a better place you have in mind.

File: src/Parsers/ExpressionListParsers.cpp [hunk 4]
```diff
@@ -1667,6 +1667,11 @@ class TrimLayer : public Layer
                 if (!mergeElement())
                     return false;
 
+                /// Skip replaceRegex if trying to TRIM an empty string to avoid exception
+                if (char_override && elements[0]->getID() == "Literal_''") {
```
Comment (alexey-milovidov) [inline src/Parsers/ExpressionListParsers.cpp:1671]: String comparison is not good (Perl-style), replace with checking the type and checking the literal value exactly.

**Review by alexey-milovidov — APPROVED** [review]

**Comment by robot-clickhouse-ci-2** [comment]
*This is an automated comment for commit 4edbf035047bf9ba4ed77ad6d5002a246bda6232 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/68455/4edbf035047bf9ba4ed77ad6d5002a246bda6232/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Docs check</td><td>Builds and tests the documentation</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/d3c9943147ad312b4df3c3c6bcdaf9d7474be2ca/docs_check.html">✅ success</a></td></tr>
<tbody>
</table>
</details>

**Pull request merged**